opal/mca/btl/iwarp/btl_iwarp.c

```diff
@@ -0,0 +1,2009 @@
+    if (NULL != local_cpc->cbm_endpoint_init) {
+        rc = local_cpc->cbm_endpoint_init(endpoint);
+        if (OPAL_SUCCESS != rc) {
+            OBJ_RELEASE(endpoint);
+            return OPAL_ERROR;
+        }
+    }
+
+    *endpoint_ptr = endpoint;
+    return OPAL_SUCCESS;
+}
+
+static int get_iwarp_btl_params(mca_btl_iwarp_module_t* iwarp_btl, int *port_cnt_ptr)
+{
+    int port_cnt = 0, rank = -1, j;
+    for(j=0; j < mca_btl_iwarp_component.ib_num_btls; j++){
+        if(mca_btl_iwarp_component.iwarp_btls[j]->port_info.subnet_id
+           == iwarp_btl->port_info.subnet_id) {
+            if(iwarp_btl == mca_btl_iwarp_component.iwarp_btls[j]) {
+                rank = port_cnt;
+            }
+            port_cnt++;
+        } else {
+            if (mca_btl_iwarp_component.allow_different_subnets) {
+                if (iwarp_btl == mca_btl_iwarp_component.iwarp_btls[j]) {
+                    rank = port_cnt;
+                }
+                port_cnt++;
+            }
+        }
+    }
+    *port_cnt_ptr = port_cnt;
+    return rank;
+}
+
+/*
+ *  add a proc to this btl module
+ *    creates an endpoint that is setup on the
+ *    first send to the endpoint
+ */
+int mca_btl_iwarp_add_procs(
+    struct mca_btl_base_module_t* btl,
+    size_t nprocs,
+    struct opal_proc_t **procs,
+    struct mca_btl_base_endpoint_t** peers,
+    opal_bitmap_t* reachable)
+{
+    mca_btl_iwarp_module_t* iwarp_btl = (mca_btl_iwarp_module_t*)btl;
+    size_t nprocs_new_loc = 0, nprocs_new = 0;
+    int i,j, rc;
+    int lcl_subnet_id_port_cnt = 0;
+    int btl_rank = 0;
+    volatile mca_btl_base_endpoint_t* endpoint;
+
+    btl_rank = get_iwarp_btl_params(iwarp_btl, &lcl_subnet_id_port_cnt);
+    if( 0 > btl_rank ){
+        return OPAL_ERR_NOT_FOUND;
+    }
+
+#if HAVE_XRC
+    if(MCA_BTL_XRC_ENABLED &&
+            NULL == mca_btl_iwarp_component.ib_addr_table.ht_table) {
+        if(OPAL_SUCCESS != opal_hash_table_init(
+                    &mca_btl_iwarp_component.ib_addr_table, nprocs)) {
+            BTL_ERROR(("XRC internal error. Failed to allocate ib_table"));
+            return OPAL_ERROR;
+        }
+    }
+#endif
+
+    rc = prepare_device_for_use (iwarp_btl->device);
+    if (OPAL_SUCCESS != rc) {
+        BTL_ERROR(("could not prepare iwarp device for use"));
+        return rc;
+    }
+
+    if (0 == iwarp_btl->num_peers) {
+        /* ensure completion queues are created before attempting to
+         * make a loop-back queue pair */
+        rc = iwarp_btl_size_queues(iwarp_btl);
+        if (OPAL_SUCCESS != rc) {
+            BTL_ERROR(("error creating cqs"));
+            return rc;
+        }
+    }
+
+    /* prepare all proc's and account them properly */
+    for (i = 0, nprocs_new_loc = 0 ; i < (int) nprocs; i++) {
+        struct opal_proc_t* proc = procs[i];
+        mca_btl_iwarp_proc_t* ib_proc;
+
+#if defined(HAVE_STRUCT_IBV_DEVICE_TRANSPORT_TYPE)
+        /* Most current iWARP adapters (June 2008) cannot handle
```

Comment:
Make having IBV_DEVICE_TRANSPORT_TYPE be required for this BTL (i.e., remove the `#if`).